### ISSUE TITLE
SD card related APIs seem to be broken on Artemis boards

### ISSUE THREAD

**Issue by idea--list** [sparkfun/mbed-os-ambiq-apollo3#36]
### Description of defect
Some days ago i closed #34 as i by getting back the size and other data of my SD card, i thought it will work. So i ported my project from the MAX32630FTHR board to Artemis Thing Plus. The core functionality of the project is datalogging on the fly. When using the Artemis board at the end of the recording session all i get is an erased, empty SD card.
While on the MAX32630FTHR i always get my files as expected.
This has led me to test Mbed's SD examples once again on the Artemis Thing Plus. It seems, the board can somehow get the size of the SD card, despite it fails to initialize/deinitialize the SDBlockDevice. Pretty strange!

#### Target(s) affected by this defect ?
Artemis Thing Plus

#### Toolchain(s) (name and version) displaying this defect ?
ARMC 6.14

#### What version of Mbed-os are you using (tag or sha) ?

Mbed OS 6.3.0

#### What version(s) of tools are you using. List all that apply (E.g. mbed-cli)
Mbed Studio V1.2.1

#### How is this defect reproduced ? 
Try this code:

```
#include "mbed.h"
#include "SDBlockDevice.h"

#if TARGET_MAX32630FTHR
#include "lib/max32630fthr/max32630fthr.h"
MAX32630FTHR pegasus(MAX32630FTHR::VIO_3V3);
#endif

Timer timer;
// Instantiate the SDBlockDevice by specifying the SPI pins connected to the SDCard
// socket. The PINS are:
//     MOSI (Master Out Slave In)
//     MISO (Master In Slave Out)
//     SCLK (Serial Clock)
//     CS (Chip Select)
SDBlockDevice sd(MBED_CONF_SD_SPI_MOSI, MBED_CONF_SD_SPI_MISO, MBED_CONF_SD_SPI_CLK, MBED_CONF_SD_SPI_CS);
uint8_t block[512] = "Hello World!\n";
int main()
{
    timer.start();
    // Call the SDBlockDevice instance initialisation method
    if (0 != sd.init()) {
        printf("Init failed \n");
        return -1;
    }
    printf("sd size: %llu\n",         sd.size());
    printf("sd read size: %llu\n",    sd.get_read_size());
    printf("sd program size: %llu\n", sd.get_program_size());
    printf("sd erase size: %llu\n",   sd.get_erase_size());
    // Set the frequency
    if (0 != sd.frequency(5000000)) {
        printf("Error setting frequency \n");
    }
    if (0 != sd.erase(0, sd.get_erase_size())) {
        printf("Error Erasing block \n");
    }
    // Write data block to the device
    if (0 == sd.program(block, 0, 512)) {
        // Read the data block from the device
        if (0 == sd.read(block, 0, 512)) {
            // Print the contents of the block
            printf("%s", block);
        }
    }
    // Call the SDBlockDevice instance de-initialisation method
    sd.deinit();
    timer.stop();
    printf("Done. It took: %d milliseconds\n", timer.read_ms());
    timer.reset();}
```

On the MAX32630FTHR it takes just 94 ms to run, while on the Artemis Thing Plus 12846 ms and on this board "Hello world!" is not prompted back in the terminal.

**Comment by idea--list** [comment]
Inside SDBlockdevice.cpp i turned on debugging like that:
```
#define SD_DBG                                   1      /*!< 1 - Enable debugging */
#define SD_CMD_TRACE                             1      /*!< 1 - Enable SD command tracing */
```
Compiled the example for both targets and this is the comparison using the same card and card slot:
![image](https://user-images.githubusercontent.com/5784071/95568339-6389ae80-0a24-11eb-9ec3-c1e8f3fdb62f.png)
On the Artemis Thing Plus things go wrong when trying to write to the card. More precisely when `SDBlockDevice::_write()` gets called from within `SDBlockDevice::program()`. As next step i commented out `if (0 == sd.program(block, 0, 512)) {` inside main.cpp, compiled&run the code. This time everything went fine, so `SDBlockDevice::read()` seems to be working fine.

I am getting closer to the root of the issue, but am not sure if i can solve it. My current assumption is that some pins' logic state might be different for some reason and that might be causing the issue. As i only have a voltmeter, i can only check the default states of the pins with an empty program.


**UPDATE**
Commented out these lines from the example:
```
if (0 == sd.program(block, 0, 512)) {
        // Read the data block from the device
        if (0 == sd.read(block, 0, 512)) {
            // Print the contents of the block
            printf("%s", block);
        }
    }
    // Call the SDBlockDevice instance de-initialisation method
    sd.deinit();
```
I assume that way both boards should show the same HI/LO states on all the pins (might be wrong).
Artemis' A5 pin has a voltage of 0,12, MAX32630FTHR shows 1,8V (not sure if that is HI or LO, but other pins have 3,3 on that board).
D9 pin of the Artemis has 0,13 that of the MAX32630FTHR 3,3V.
However these 2 pins should not be involved in SPI communication with the SD card, so this might not be relevant to the issue.
I can not go further as i do not have a logic analyzer.

**Comment by idea--list** [comment]
Today i wrote a very simple, 2-channel logic analyzer code (using Mbed's AnalogIn API), that feeds Arduino's serial plotter (the only feature of Arduino why i have it installed 😉 )
Both the MAX32630FTHR and the Artemis Thing Plus were running the same SD example code i posted in the first post.
The logic analyzer code was running on a second MAX32630FTHR, and made 2 recordings for each board. One recorded MOSI and MISO, another captured MOSI and CS signals.

Results of the MAX32630FTHR:
![MAX32630FTHR](https://user-images.githubusercontent.com/5784071/95652925-dd896880-0af4-11eb-8941-6673f528ec66.jpg)

Results of the Artemis Thing Plus:
![ARTEMIS](https://user-images.githubusercontent.com/5784071/95652948-0d387080-0af5-11eb-8007-c711ec394daa.jpg)

One can clearly see that the MOSI signal is quite different on the Artemis board.
I just wonder why this might happen considering the codebase is the same on both boards?
Why does the MOSI pin behave different on the Artemis and how to overcome it?

**Comment by idea--list** [comment]
Ok it seems i might have found something:
- the example code fails when SDBlockDevice API's `program()` function gets called.
- inside that function a block write is requested (line 506) relying on the `_write()` helper function also defined in SDBlockDevice.cpp
- inside the `_write()` function the `write()` function (defined for each target differently! in spi_api.c inside the targets/ folder) gets called several times: on lines 983, 986, 997, 998. Even on line 998 both boards get the same response from the SD card. On line 1002 however `response = _spi.write(SPI_FILL_CHAR);` will result 0x5E (dec 229) on the MAX32630FTHR, but on the Artemis the result here is different: 0x00 (dec 0). Because of line 1009, `_write()` will return a wrong value to line 506.
- digging one layer deeper it turns out that `write()` defined in SPI.cpp relies on `spi_master_write()` and `spi_master_block_write()` functions, which are defined for each target differently within spi_api.c file found in the targets/ folder.

For the Artemis Thing Plus i can see this:
```
int spi_master_write(spi_t *obj, int value)
{
    uint32_t rxval = 0;
    spi_master_block_write(obj, (const char *)&value, 1, (char *)&rxval, 1, 0x00);
    return rxval;
}
```
Might be wrong but it seems it hardcodes 0x00 as fill character despite in case of SD cards that should be 0xFF.
Also `spi_master_block_write()` seems to hard code the fill character to 0x00 (line 180).

- Furthermore also the implementation of `spi_master_block_write()` seems to be buggy for the Artemis board:
line 506 in SDBlockDevice.cpp passes arguments to spi_master_block_write() (via line 986 of SDBlockDevice.cpp).
Originally the code tries to write 512 blocks, but the Artemis board can not finish writing 512 blocks and the `_wait_ready()` function called on line 1005 will simply shut down the process with a "Card not ready yet" message.
I found the Artemis implementation of `spi_master_block_write()` can cope with 32 blocks within 1 second. With 32 blocks the process will not time out, but will still fail and we get the "Single Block Write failed" message. For more blocks than that even the deafult 10000 msec is not enough. (Actually i also tried to set an extreme long timeout value of 600 secs in mbed_app.json, but even that much time is not enough to finish writing 512 blocks).

So i got the feeling both `spi_master_write()` and `spi_master_block_write()` functions should be reviewed and reimplemented for the Artemis board.

**Comment by idea--list** [comment]
When i manually limit the bumber of blocks to 32 on line 506 in SDBlockDevice.cpp, then both boards have exactly the same responses after each spi.writes. That is promising, maybe i just need to find out how to allow the Artemis board working with 512 blocks...

**Comment by idea--list** [comment]
Looked more into the details of spi_api.c.
Remember Mbed's SDBlockDevice.cpp  contains this: `_spi.write((char *)buffer, length, NULL, 0);` where length is set to be 512.
Artemis board's spi_api.c file contains this on line 180: 
```
if (!Rw || (write_fill == 0x00)) { // when transmitting (w) or reading with a zero fill just use a simplex transfer
            status = am_hal_iom_blocking_transfer(obj->spi.iom_obj.iom.handle, &xfer);
```
So the SPI communication is not full duplex when line 986 in SDBlockDevice.cpp is executed.
~~To overcome this, i modified the above condition on line 180 to:~~
`if ( (write_fill == 0x00)) {`
~~In this case the block between lines 187-197 gets executed with a full duplex blocking transfer.
Now writing those 512 blocks does not result a time out (No "card not ready yet" message any more)~~

~~However it still fails with "Single Block Write failed: 0x1f" message~~

**Comment by idea--list** [comment]
@oclyke 
Have a working solution for the issue. The SD functionality works both with Mbed's SD-related examples and even in my project.
My question is where should i make a pull request? Here in this repo or rather in the official Mbed repo over [here](https://github.com/ARMmbed/mbed-os)?
(Could not test it extensively as i only have the Artemis Thing Plus, and also did not try other SPI devices than SD-cards)

**Comment by oclyke** [comment]
Thanks @idea--list, we'd certainly like to see a PR.

The right person to ask about that is @Wenn0101. With that said, any thoughts?

**Comment by idea--list** [comment]
@Wenn0101 Can you please let me know in which repo shall i put the PR?
I mean if i had all the Artemis boards and could also test several other SPI devices without any regression then it would be an easy decision. So basically the question is if you guys could review & test the fix if i place it here?
I mean Mbed team might not have your boards to test and they only test about a 6-8 different boards nothing more... so testing would be important to prevent any unexpected regression.

**Comment by idea--list** [comment]
Just made PR #39 which fixes the issue that occurs when trying to write to an SD card with the Artemis MCU.

**Comment by idea--list** [comment]
Has been fixed in mBed OS v6.5
